feat(logger): add clear_state method (#5956)
awslabs/aws-lambda-powertools-python: docs/core/logger.md

#### Clearing all state

##### Decorator with clear_state

Logger is commonly initialized in the global scope. Due to [Lambda Execution Context reuse](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-context.html){target="_blank"}, this means that custom keys can be persisted across invocations. If you want all custom keys to be deleted, you can use `clear_state=True` param in `inject_lambda_context` decorator.

???+ tip "Tip: When is this useful?"
    It is useful when you add multiple custom keys conditionally, instead of setting a default `None` value if not present. Any key with `None` value is automatically removed by Logger.

???+ danger "Danger: This can have unintended side effects if you use Layers"

>     Lambda Layers code is imported before the Lambda handler. When a Lambda function starts, it first imports and executes all code in the Layers (including any global scope code) before proceeding to the function's own code.

Reviewer: SUPER NICE! 🚀

    --8<-- "examples/logger/src/clear_state_event_two.json"
    ```

##### clear_state method

> `clear_state()` is a method you can call explicitly within your code to clear appended keys at any point during the execution of a single Lambda invocation.

Reviewer: ```suggestion
You can call `clear_state()` as a method explicitly within your code to clear appended keys at any point during the execution of your Lambda invocation.
```

It's better to use active voice.

Author: Perfect!! Thanks, changed.

> This allows for more granular control over the logger's state within a single function execution, enabling you to reset the logger to its initial state before specific logging operations or at the end of certain processes within the same Lambda run.

Reviewer: I think we can remove this paragraph. 

> state within a single function execution,
> of certain processes within the same Lambda run.

These statements may be invalid due to Lambda's execution model and state persisting between invocations, so this means that if you add a key in one execution, it may persist between executions:

```python
from aws_lambda_powertools import Logger

logger = Logger()

def lambda_handler(event, context):
    logger.info("No keys")
    logger.append_keys(mykey="key")
    logger.info("With keys")
```

First execution will log `no keys` without mykey
First execution will log `no keys` with mykey

Author: Removed the second paragraph, and left the first one with the modification proposed earlier.
